src/passes/OptimizeInstructions.cpp

```diff
@@ -1882,17 +1882,33 @@ struct OptimizeInstructions
 
       // This local.set of a struct.new looks good. Find struct.sets after it
       // to optimize.
-      for (Index j = i + 1; j < list.size(); j++) {
+      Index localSetIndex = i;
+      for (Index j = localSetIndex + 1; j < list.size(); j++) {
         auto* structSet = list[j]->dynCast<StructSet>();
+        // Any time the pattern no longer matches, we try to push the
+        // struct.new further down but if it is not possible we stop
+        // optimizing possible struct.sets for this struct.new.
+
         if (!structSet) {
-          // Any time the pattern no longer matches, stop optimizing possible
-          // struct.sets for this struct.new.
+          if (trySwap(list, localSetIndex, j)) {
+            // Update the index an continue to try again.
+            localSetIndex = j;
+            continue;
+          }
           break;
         }
+
         auto* localGet = structSet->ref->dynCast<LocalGet>();
         if (!localGet || localGet->index != localSet->index) {
+          if (trySwap(list, localSetIndex, j)) {
+            // Update the index an continue to try again.
+            localSetIndex = j;
+            continue;
+          }
```

Comment:
Using a lambda could help avoid the duplication here. The only thing a lambda can't do is the continue. So you could have
```cpp
auto doTrySwap = [&]() {
  if (trySwap(list, localSetIndex, j)) {
    localSetIndex = j;
    return true;
  }
  return false;
};
```
And then uses would be
```cpp
if (doTrySwap()) {
  continue;
}
```

However, maybe the duplication can be fixed another way, like setting a boolean and then doing the `trySwap` near the end, if the boolean was set?

Reply:
Restructured the code to avoid duplication, but not sure if it is easier to read.